### DISCUSSION TITLE
WMS/WMTS Tile Service Integration?

### DISCUSSION THREAD

**Discussion by jfuehner** [heremaps/harp.gl#2277] [Q&A]
Hi! I am new to harp and am looking at using it for integrating with a WMS or WMTS tile service.  I thought perhaps this [example ](https://github.com/heremaps/harp.gl/blob/9fd4eb5cca975038e9b7f6f46506fd48c4577b94/%40here/harp-examples/src/datasource_webtile.ts) was the correct route. 

I can see tile network requests going out to the server and tiles being returned. However, nothing is rendered to the screen. The screen appears completely black? Additionally, the `tileKey`'s `level `property for some reason starts at 5 when I would expect 0?

Any assistance would be great!

**Comment by nzjony** [comment]
As far as I'm aware, we don't support WMS or WMTS, though probably you could add support.

That example doesn't exist anymore, if you search on the master branch it is gone, do you mean instead: https://www.harp.gl/docs/master/examples/#datasource_here-webtile.html ?

I guess you need to checkout out the latest master and retry?

**Reply by jfuehner** [reply]
@nzjony I was looking at this example https://www.harp.gl/docs/master/examples/#datasource_custom-texture-tile.html. 

I got the impression the `getTexture` function could be used to call an external url to retrieve a tile at the `xyz` location for say a WMTS service?

**Reply by nzjony** [reply]
Yes, that should work.

Here is the class which makes requests to the Here tile services: https://github.com/heremaps/harp.gl/blob/b677b8717f06ceec696a22a7ea7bd38f34653b02/%40here/harp-webtile-datasource/lib/HereWebTileDataSource.ts#L177

Which internally uses this class to download the tiles: https://github.com/heremaps/harp.gl/blob/b677b8717f06ceec696a22a7ea7bd38f34653b02/%40here/harp-mapview/lib/TextureLoader.ts#L16

It should be possible to take that as a starting point and just change the URLs / authentication mechanism. All the best!

**Reply by jfuehner** [reply]
@nzjony thats super helpful thanks!

Is harp able to render tiles stored in CRS 4326? Any thoughts as to why the tileKey level was requesting level 5 for the first set of tiles?

**Reply by nzjony** [reply]
Level 5 is just some arbitrary default, if you don't set anything, see: https://github.com/heremaps/harp.gl/blob/e5bbad2d042f1976f814b029e1115a5a661d566e/%40here/harp-mapview/lib/MapView.ts#L610

Please forgive my ignorance, but what do you mean by:
> render tiles stored in CRS 4326 ?

We can render GeoJson, which describes data in lat / Lon pairs ok.

**Reply by jfuehner** [reply]
@nzjony CRS 4326 is the EPSG number for World/Global Geodetic
https://en.m.wikipedia.org/wiki/World_Geodetic_System the WGS84 reference system.

vs say EPSG 3857 which is the Web Mercator Projection

https://en.m.wikipedia.org/wiki/Web_Mercator_projection

**Reply by nzjony** [reply]
@jfuehner. Ok, so when you ask whether we can render tiles in the WGS84 system, do you mean render the points as a globe (i.e. a spherical projection as opposed to a planar projection)?

**Reply by jfuehner** [reply]
@nzjony I was able to use the getTexture example you pointed me to render some tiles from my tile server. However, the entire earth is not displaying. This is the test arc server source I am using...

https://services.arcgisonline.com/arcgis/rest/services/ESRI_Imagery_World_2D/MapServer/tile/{z}/{x}/{y}

North/South America and Africa appear duplicated across the screen. None of the other tiles for Asia are retrieved and displayed.

What am I missing?

**Reply by nzjony** [reply]
What projection are you using? And what projection are the images in?

Do you have some repo I can try and debug? Or a screenshot at least?

**Reply by jfuehner** [reply]
@nzjony sorry about that, I should have provided more details...

The ESRI map service is projected in the spatial reference [EPSG:4326](https://epsg.io/4326). You can see the service description by browsing to:

https://services.arcgisonline.com/arcgis/rest/services/ESRI_Imagery_World_2D/MapServer

I mocked up a StackBlitz [example](https://vue-ubsamc.stackblitz.io) of what I am seeing. In the example I am trying to use harp with deck.gl.

**Reply by nzjony** [reply]
@jfuehner , thanks for the update. It seems you need to find the correct tiling scheme, if I'm not mistaken, you can pass this into the DataSource constructor. Basically, it describes how the tiles are subdivided. It seems that it has some half quad tree subdivision scheme? Like: https://github.com/heremaps/harp.gl/blob/master/@here/harp-geoutils/lib/tiling/HalfQuadTreeSubdivisionScheme.ts

Or download the harp.gl code and search for TilingScheme.

I hope that helps.

**Reply by jfuehner** [reply]
@nzjony I tried simply passing the `tilingScheme` property as a part of the data source construction. However, that didn't seem to make a difference...

https://codesandbox.io/s/serverless-wildflower-7orbf

```
    const webTileDataSource = new WebTileDataSource({
      tilingScheme: halfQuadTreeSubdivisionScheme,
      dataProvider: {
        getTexture: context.getTexture,
      },
    });
```

Looking at the `WebTileDataSource` I only see a _getter_ for the tiling scheme. In looking at the harp website [CustomDataSource](https://www.harp.gl/docs/master/examples/codebrowser.html?src=src%2Fdatasource_custom.ts) example I see where a `tilingScheme `is being passed to the constructor of a custom data source. 

I'm wondering if perhaps I can't specify a `tilingScheme ` with the `WebTileDataSource `but instead need to create a custom data source?

**Reply by nzjony** [reply]
Sorry, I was wrong, it seems that you need to override it:

I.e. create a class that extends from `WebTileDataSource` and override `getTilingScheme`

**Comment by jfuehner** [comment]
@nzjony hmm... extending the `WebTileDataSource `/ `getTilingScheme` seems to get me a little farther. However, the `halfQuadTreeSubdivisionScheme` does not appear to expose/implement the `getGeoBox` _(See error screen capture)_ function.

https://codesandbox.io/s/serverless-wildflower-7orbf

![image](https://user-images.githubusercontent.com/1843941/129575735-759afbb9-70c5-448e-8cfe-7725bcdf3518.png)

Do I need to implement my own titling scheme class?

**Reply by jfuehner** [reply]
@nzjony a little trial and error and was able to get the extended `WebTileDataSource `class to start rendering tiles...

```
class CustomWebTileDataSource extends WebTileDataSource {
  getTilingScheme() {
    return new TilingScheme(halfQuadTreeSubdivisionScheme, webMercatorProjection);
  }
}
```

However, they still appear incorrect. I assume it is due to the projection I've passed. I tried the other available projections I have found in the `harp-geoutils/projection` package. None seem to project the tiles correctly...

**Reply by nzjony** [reply]
Screen shot?

Yes, maybe you need to implement your own tiling scheme class to fit the way these are cut. I was hoping one of the existing ones would work... but obviously not.

**Reply by jfuehner** [reply]
@nzjony 

Looking at the projection classes they seem a bit out of my depth.  Ugh... Was hoping this was going to be easier.

**Reply by nzjony** [reply]
@jfuehner , I don't think the issue is the projection, I think it is the tiling scheme. 

At least the images seem to be in equirectangular projection? You just need to know how to piece the tiles together.

Please send through another update and maybe I can look at it tomorrow.

**Comment by jfuehner** [comment]
@nzjony here is the CodeSandbox I have been working with...

https://codesandbox.io/s/serverless-wildflower-7orbf

OpenLayers actually has an [example ](https://openlayers.org/en/latest/examples/xyz-esri-4326-512.html)on using an ArcGIS tile service endpoint in EPSG:4326 projection. Looks like they set the `maxResolution: 180 / 512`. Not sure how this would translate to harp.

https://openlayers.org/en/latest/examples/xyz-esri-4326-512.html

If I try using the `equirectangularProjection` it looks like this. 

![image](https://user-images.githubusercontent.com/1843941/129594167-8be21dc9-e01f-4c06-b139-f25c5271ac94.png)

When I zoom in the tiles go all out of whack...

![image](https://user-images.githubusercontent.com/1843941/129594257-3e728bda-ea89-4387-9a97-5b467184d4cf.png)

The `webMercatorProjection` seems to be the closest.

**Reply by nzjony** [reply]
I checked this: https://gis.stackexchange.com/questions/387517/what-is-meant-with-epsg4326-projection

EPSG4326 isn't actually a projection, it is just a reference system. I guess that is why I was confused. But it seems if you just put the numbers onto a grid, you get something that looks Equirectangular, at least that is what the images look like, the tiles don't look like the mercator projection.

The screenshot you shared seems good, which tiling scheme do you use? I Think you just need to tweak the way the tiles are subdivided.

**Reply by nzjony** [reply]
I.e. at level 0, how many tiles are there in the X direction and in the Y direction?
At level 1, how many are there


And so on...

And make a class which captures this

I.e.

```
class CustomSubdivision implements SubdivisionScheme {
    getSubdivisionX(): number {
        return <what you found out above>;
    }

    getSubdivisionY(level: number): number {
        return <what you found out above>;
    }

    getLevelDimensionX(level: number): number {
         return <what you found out above>;
    }

    getLevelDimensionY(level: number): number {
        return <what you found out above>;
    }
}

**Reply by jfuehner** [reply]
@nzjony I created a custom implementation of the subdivision scheme. However, modifying the x/y subdivisions do not seem to be having the desired impact.

You can see a working copy here...

https://codesandbox.io/s/serverless-wildflower-7orbf

```
export class CustomSubdivisionScheme {
  getSubdivisionX() {
    return 2;
  }

  getSubdivisionY() {
    return 2;
  }

  getLevelDimensionX(level) {
    return 1 << level;
  }

  getLevelDimensionY(level) {
    return 1 << level;
  }
}

class CustomWebTileDataSource extends WebTileDataSource {
  getTilingScheme() {
    return new TilingScheme(
      new CustomSubdivisionScheme(),
      equirectangularProjection
    );
  }
}
```

**Reply by nzjony** [reply]
@jfuehner , according to: https://services.arcgisonline.com/arcgis/rest/services/ESRI_Imagery_World_2D/MapServer

The first level 0 has only 2 images, so getSubdivisionY should be like: 

```
getSubdivisionY(level: number): number {
        return level === 0 ? 1 : 2;
    }

```

**Reply by nzjony** [reply]
Or... something like that, maybe it is subdivision X.
